### ISSUE TITLE
Goal and user cases for next step

### ISSUE THREAD

**Issue by xhdong-umd** [ctmm-initiative/ddim#1]
@chfleming , @NoonanM ,  can you provide some example user cases that you would like to be implemented for next step of the project?

I think we discussed about retrieving remote sensing data through `rstools`, also some typical metrics needed for annotation. I'm not familiar with them yet, so some examples will be helpful on details.

**Comment by chfleming** [comment]
I think the simplest example that would be useful to users would be to have the capability to plot data and distributions atop NDVI or EVI layers. These are satellite data products that corresponds closely to vegetation. NDVI/EVI data comes in large tiles estimated from a ~15 day window and in a weird projection. So plotting would involve determining the relevant time frames, the relevant spatial extent, and matching the projections. `RStoolbox ` should make a lot of that easier.

**Comment by chfleming** [comment]
@xhdong-umd @jmcalabrese I'm starting to work on this material now, so here is what I'm thinking for doing resource selection functions and animations.

1. After setting up the global variables (user name, password, directories, ...), I'd like to have a function like 

`get.envdata("NDVI",ext=NULL,buffer=0,res.space=1,res.time=1,...)`

where the first argument would be the type of data requested,
the `ext` argument would be the base extent (or a `telemetry`, `ctmm`, or `UD` object to run through `extent()`),
the `buffer` would increase the spatial extent from the center by that factor,
`res.space=1` would correspond to the natural spatial resolution of the data, so 250 (500?) meters for the MODIS data,
`res.time=1` would correspond to the natural temporal resolution of the data, so 16 days for the MODIS data.
We don't have to support everything at first.

I need to update the `extent()` function to store projection and temporal information, so for the moment just hold off on that argument.

2. The final object should be an array gridded in the projection of `ext`, though I understand that it needs to be long-lat for the request. So, first, construct the grid in the projected coordinate system of `ext`, then inverse-project to long-lat (you can use `ctmm:::project` for that) for the actual request.

3. The final object should be a 3D array with dimensions of time-x-y and also supplied with grid specification vectors `t` (which are `as.numeric` compatible with `telemetry$t`), `x`, `y`, as well as time epoch and projection information. If this doesn't fit conveniently within the `raster` schema (bricks or stacks?), we can make a new object class.

**Comment by xhdong-umd** [comment]
Thanks @chfleming for the information. I'm having some road blocks in submitting grid request to movebank ENV-Data, and I just sent an email to ask again with more details.

Once we are more clear about what can be done then we can discuss on the details of this.

**Comment by chfleming** [comment]
As a more long term goal, It would be good to know and utilize

1. The timing of the environmental data, so sample the layers at those times
2. The native projection and grid of the environmental data, to get the native grid

(1) is more important than (2), at least for NDVI, but in both cases, having the native sample would allow to avoid interpolating the data twice, which is going to oversmooth the data.

**Comment by xhdong-umd** [comment]
I now have the backend functions mostly in shape and tested some examples. I'll start to look at these use cases.

First I will try to get some result and report back what I got. The result could be grid csv using track mode or png/geotiff/kmz using grid mode. Will tiff work for the user cases? Looks like we may need to get grid csv which is most flexible, but it will need to upload a bigger csv compare to grid mode (which only need a bounding box and resolution).

**Comment by xhdong-umd** [comment]
@chfleming , some questions:
- `ctmm::extent` on buffalo data will get bounding box of x,y instead of lon/lat. Right now I merge all telemetry object data then create a simple feature points data.frame, then get the bbox. This way I got all the information including proj saved, also easy to convert if needed.
- As for projection, the backend data input default projection in WGS84, to customize you need to input epsg code. The instruction said something like this

> number of pixels (x,y,per tile): The resolution you want for the output. Typically choose it to be in roughly the same
> range as the resolution of the original product because we are not integrating over an area, but interpolating to the
> grid points nearest to your pixel. Let us know if you need help figuring this out. **Data are always re-projected**.

- one way is we use original data coordinates and customize the projection in input parameter, but that need an epsg code. I think there is no automatic way to always get an epsg code from proj string, right?
- the easier way might be we always project original data into WGS84. We may not be able to avoid re-projection according to how the backend worked.

**Comment by chfleming** [comment]
I just pushed updates to extent that also return long/lat and other extents. The first two columns are still x,y, but others can be present if present in the data. This could break code where extent output is assumed to only have two columns and you will need to substitute `extent(...)` with `extent(...)[,c('x','y')]`.

**Comment by chfleming** [comment]
PROJ4 projections do not necessary have EPSG codes. Custom PROJ4 projections, like `ctmm` uses, definitely won't.

I'd like to get whatever is closest to the original data, even if that is WGS84.

**Comment by xhdong-umd** [comment]
Which data is the original data you mentioned? The animal tracking data, or MODIS product data?
- In grid mode we need to specify a bounding box with projection, I think a re-projection will happen anyway unless the animal data is in exactly same projection with MODIS data.
- For input, I think we either project to something common like WGS84 or to the MODIS projection, which seemed to be `the Sinusoidal (SIN) grid projection (equal area projection)`.
- Then I need to check the returned result projection.

And, if I need to reproject data, how can I use extent? I knew I can take the location points and do some projection, then should I call extent on points again? I noticed that extent on a list of telemetry objects is different from the extent on all these locations in single data.frame and treated like single individual.

Is there some recommended way to reproject telemetry obj locations and still keep in telemetry obj format?

**Comment by chfleming** [comment]
Sorry, the MODIS product, as a long-term goal I'd like that as close to the original as possible, and then I can iterate from that. The last time I downloaded MODIS data, fine-scale data were in a very inconvenient sinusoidal projection and coarse-scale data were in long-lat. As a long-term goal, if you can get something close to the original product downloaded, then I will focus on the projections. I'd like to implement something non-destructive that would keep a copy of the original in the object.

But just to get things started, 16-day interval, 500 meter resolution MODIS data in a convenient coordinate system like WGS84 is fine. I can code the projection.

I'm a little lost on the last part, but telemetry data are pretty easy to re-project:
`projection(DATA) <- PROJ4STRING`
and to make a local projection that keeps North up (so to have a similar x-y and long-lat extents):
`projection(DATA) <- median(DATA)`

How do the extents differ? It should be that the extent of a list of objects is just the union of the extents.

Have you looked at the MODIS package? https://cran.r-project.org/web/packages/MODIS/index.html
Is this useful?

**Comment by xhdong-umd** [comment]
When level is not 1, extent on each individual then union is different from treat all data as if from single individual.

I took a brief look at the MODIS package, mostly download and process the data directly. We will let backend to get and process the MODIS data so will not need this. There is a function to create time intervals, I need to check how movebank deal with our provided time stamps.

**Comment by chfleming** [comment]
`level=1` is default for telemetry objects, though...

**Comment by xhdong-umd** [comment]
I'm trying to project data to WGS84 but failed
```r
> proj4 <- sp::CRS("+init=epsg:4326")
> proj4
CRS arguments: +proj=longlat +datum=WGS84 +no_defs 
> # take simple string, not crs object
> ctmm::projection(tele_list) <- proj4
Error in validate.projection(proj) : 
  A projected coordinate system must be specified.
```
I tried proj4 as simple string and crs object, both doesn't work.

Besides, extent can work on UD object, ctmm object, do you think we will use their extent for getting environment data? If so, I better always use extent, then I will always change projection on telemetry object then get extent. Otherwise I can just put out data points and change projection/get bounding box.

**Comment by chfleming** [comment]
WGS84 is the default datum and long-lat are already in the telemetry object. You don't need to reproject to extract those. I prevent users from also setting x-y to be long-lat because then the numbers are meaningless and the projection isn't locally flat and 1:1. Otherwise users will calculate speeds and areas in units of mismatched degrees.

There is a very good chance that the UD extents will get used to get environmental data. If the environmental data is fundamentally in long-lat and we use the UD extent, then I will write some code to calculate long-lat extents from the UD.

**Comment by NoonanM** [comment]
I've also noticed that in older versions of R (pre 4.0) this works just fine:
```
ud <- raster(UD)
ud <- projectRaster(ud, crs=crs("+init=epsg:4326"))
```

But in the most recent version of R the exact same code fails. A lot of packages might be struggling to keep up with the changes that came with 4.0

**Comment by xhdong-umd** [comment]
Does UD have a time attribute? At least for telemetry objects we may need to get the time range of them. It's easy for me to get it from telemetry data, but I'm not sure about other objects like UD.

**Comment by xhdong-umd** [comment]
Movebank will always project input data to product data (modis projection), and will interpolate values on space and time dimension. 

- On projection, we can reproject locations to MODIS projection by ourselves and provide it with customized projection. Or we can just provide original data, since movebank will do the projection anyway, and it will be same step. The only thing that can result to some loss is if we reproject location to something, then movebank reproject it again. Ideally we should just provide the input data without reprojection and provide the proj string. However movebank probably cannot recognize all projections.

  For now I think just reproject input data to WGS84 is simple and should work in many cases. For specific cases there will not be an good automatic solution.

- On time, say we have the time range of movement data, the easier way is just give the start (product have a starting time, that's another limiting factor), end, and certain intervals (determined by resolution you want) and the values will be interpolated. I don't think we need to match original time intervals from product (which will be quite specific and need to be maintained for product updates), right?

See these lines from `The environmental-data automated track annotation (Env-DATA) system: linking animal tracks with environmental data`, [Dodge et al. Movement Ecology 2013, 1:3](http://www.movementecologyjournal.com/content/1/1/3)

> Data will be interpolated if the
> requested resolution is higher than the native resolution
> of the dataset.
> 
> When
> a set of points is submitted for annotation by the Env-
> DATA system, each point is transformed to the coordinate
> reference system of the annotation variable using
> proper projection techniques. A set of neighboring grid
> points is then determined and the names of the files
> containing data for these grid points are identified.
> Once all necessary data sources are locally available, the
> environmental data are interpolated along all trajectory
> points.
> 
> Prior to the interpolation, the trajectories are first
> transformed to the native grid of the environmental
> variables (e.g. **the Sinusoidal grid for MODIS** or Lambert
> Conformal for NARR), if required. After the transformation,
> for each point along a trajectory, the interpolation
> first is applied in space, then in time. For each trajectory
> point pi (xi, yi, ti,), four adjacent pixels (or more if necessary)
> are located in the global grid of environmental data
> in space (Figure 4a) at two temporal timestamps before
> and after ti (Figure 4b). The values of the four neighbor
> pixels (i.e. v1…v4 and v1’…v4’) at two timestamps t and t’
> are then extracted and used to compute an interpolated
> value for the trajectory point pi.
> 
> The system allows three types of interpolation in
> space: nearest neighbor, bilinear, and inverse weighted
> distance [45,46], and two types in time: nearest neighbor
> and inverse weighted distance. The user can apply different
> methods of interpolation to space and to time. However,
> the available interpolation strategy differs according
> to the type of data. For instance, for categorical data,
> such as land-use type, only the nearest neighbor
> interpolation can be applied, whereas for continuousnumeric
> data types either a nearest neighbor, bilinear
> (in regular grids), or inverse weighted distance interpolation
> can be selected based on the resolution of data [45,46].
> Spatial distances used by the nearest neighbor and inverse
> weighted distance interpolations are calculated as great
> circle distances on the WGS84 ellipsoid.
>

**Comment by chfleming** [comment]
UD objects do not currently have any time information.

Yes, WGS84 is good enough for now, though down the road, something closer to the original product may be highly preferred. With the calculations involved, beyond the locations of the tracking data, you also have to calculate a surface integral. There are two ways to do that: One is to integrate over the grid and another is to sample randomly (which is what most people do). Fetching huge random samples from MoveBank would not be nice, so they would end up being interpolated twice.

For time, I feel the same way. Getting the resolution correct (16 days?) is good enough for now, but down the road it would be useful to have something closer to the original product. It's pretty regular, so if not maintained, the worse thing that could happen is that the times could get out of sync, putting us back to just having the correct resolution.

**Comment by xhdong-umd** [comment]
Now I have some scripts that can get the NDVI and EVI with only information provided in the API of `get.envdata("NDVI",ext=NULL,buffer=0,res.space=1,res.time=1,...)`. Of course the script just submit the request and user need to wait for the result. Then there is a function to download and unzip the results.

Quite some details have to be inferred from experiments. 
- png will have a nice and easy to check picture, but I think geotiff have more metadata and should be preferred, and it's more common to use tiff instead of png in geo related work. kmz just have a little bit more metadata, in the end I think we will just use geotiff.
- tiff have the raw NDVI values, so user have to do the color mapping by themselves. There is no standard color mapping for NDVI/EDI other than people usually map high vegetation as green. The png result looks good but I didn't find how to get that color map.
- I calculated resolution as 250m and 16 days. Specifying higher resolution doesn't really improve result space resolution.

NDVI result for buffalo (I need to add a little bit buffer to avoid points falling on edge)
![2020-12-27_163046](https://user-images.githubusercontent.com/25039897/103180135-f6d2c300-4860-11eb-971a-309af0ba0776.png)

EVI result for buffalo

![2020-12-27_163105](https://user-images.githubusercontent.com/25039897/103180139-fc300d80-4860-11eb-81b1-37cde1d633e2.png)

Next I will convert the geotiff to the array in the format you wanted.

**Comment by chfleming** [comment]
Great! I will start working on calculating the long-lat extent of UD objects.

**Comment by chfleming** [comment]
I added a `complete` argument to `extent.UD` so that it would calculate long-lat extents in addition to x-y extents.

**Comment by xhdong-umd** [comment]
I found MODISTools have some function to get the date in certain product. So in general given a time period, we can only request the product dates in that time range.

**Comment by xhdong-umd** [comment]
Movebank require to split result into tiles if the grid is too big. In that case the result will have multiple tiles for same timestamp. Should we assemble them into single tile so that one layer map to single timestamp and full space range? I think this is needed otherwise it's hard to process later.

**Comment by chfleming** [comment]
Definitely. This is all about making something easy for people to use.

**Comment by xhdong-umd** [comment]
I merged tiles and put each snapshot on timestamp as a raster layer, combined as a raster stack. The object have x, y, and layer in time. I put time in Z slot so you can use `raster::setZ/getZ` to access it. There are extent and project information in the raster stack.

Is this what you want as final object? I'm not sure the grid spec t you mentioned(is it bounding box or timestamps?)
(the names slot was cut short because full name were too long)
```r
> rt
class      : RasterStack 
dimensions : 4503, 148, 666444, 43  (nrow, ncol, ncell, nlayers)
resolution : 0.002321137, 0.002257208  (x, y)
extent     : 31.64416, 31.98769, -25.16421, -15  (xmin, xmax, ymin, ymax)
crs        : +proj=longlat +datum=WGS84 +no_defs 
names      : X250m_16_//8000000000, X250m_16_//6000000000, X250m_16_//2000000000, X250m_16_//7000000000, X250m_16_//3000000000, X250m_16_//9000000000, X250m_16_//5000000000, X250m_16_//0000000000, X250m_16_//6000000000, X250m_16_//2000000000, X250m_16_//8000000000, X250m_16_//3000000000, X250m_16_//9000000000, X250m_16_//4000000000, X250m_16_//0000000000, ... 
min values :            -0.1981319,            -0.1991000,            -0.1988000,            -0.1947997,            -0.1924860,            -0.1861000,            -0.1990000,            -0.1956214,            -0.1948000,            -0.1931000,            -0.1848000,            -0.1750490,            -0.1700785,            -0.1896524,            -0.1832732, ... 
max values :             0.7979858,             0.7273996,             0.7515214,             0.7242462,             0.6923314,             0.6847857,             0.6611980,             0.6597658,             0.6398827,             0.7026723,             0.7320547,             0.6879522,             0.6380293,             0.5648009,             0.5858998, ... 
time        : 2005-02-18 - 2006-12-19 (range)

```

**Comment by xhdong-umd** [comment]
When adjusting a bounding box by buffer, can we always assume `xmin < xmax` ?  I think a box can have the west edge on longitude 178 and east edge on -178 and cross the International Date Line(if we write this as reversed (xmin = -178, xmax = 178) that should mean the other side and covering most of globe.), like 

```r
xmin ymin xmax ymax 
 178   80 -178   81 
```

Then the calculation need to deal with properly. I found a bounding box calculation function from tmap which cannot deal with this.

```r
 >b1 <- tmaptools::bb(rbind(c(178, -178), c(60, 61)))
>tmaptools::bb(b1, ext = 1.1)
   xmin    ymin    xmax    ymax 
 195.80   59.95 -195.80   61.05 

```

Am I overthinking this?

**Comment by chfleming** [comment]
Good question. That can happen with long-lat. I've been avoiding the issue by always working in a locally flat projection, as projected from ellipsoidal coordinates (not long-lat). I will need to update my extent() calculations to take this into account.

**Comment by xhdong-umd** [comment]
I have the code for following functions now, except for some not implemented points:

- a function to generate a request from telemetry data(since we have to submit the request then wait the result, we cannot get result in one function call), something like `request_envdata <- function(data_type, ext, time_range, buffer = 0, res_space = 1, res_time = 1, ...)`
    data_type can be `NDVI`, `EVI`
    time_range is waiting for ext or other function to return time range of data. Right now I can calculate it by myself
    buffer is not implemented yet because of the special case mentioned. I'll implement it after the ext update
    the code didn't change projection, just use it directly as WGS84. We can try product native projection later.
- with the request generated, I have function to submit it, then another function to download result once result is ready. I have another function to unzip the result, merge tiles if it was split in tiles, pack into raster stack, and put time information of each snapshot into Z dimension of raster.

**Comment by chfleming** [comment]
I've updated my `extent` code to cross the anti-meridian. My code for longitude medians and quantiles is in `circle.R` if you happen to need it. I first tried a circular statistics package's median but the results weren't always coming out right. The only assumption I made is that data were constrained to <= 180 degree range of longitudes (so no global outliers). This allowed me to trivialize the calculation. I think the more general case requires optimization.

**Comment by chfleming** [comment]
Looks like I didn't include timestamps in the `extent.telemetry`. I'll get on that.

**Comment by chfleming** [comment]
`extent.telemetry` produces timestamp extents now.

**Comment by xhdong-umd** [comment]
I updated my function with the new extent function. I also made the buffer parameter support anti-meridian case. However I'm not sure how Movebank deal with that case. They require a origin + shape
- origin (x,y): Enter the coordinates of the northwest corner of the area of interest (bounding box) in decimal degrees(long,lat)
- shape (x,y): Enter the distance that the bounding box extends from the northwest corner in decimal degrees (long,lat; x should be positive and y should be negative)

So they assume xmax > xmin but in anti-meridian case west point will have value in 0 -180, east point have -180 - 0. I guess the bounding box will have some problem. I'm not sure how rare will this happen. I probably need to ask them about this.

**Comment by xhdong-umd** [comment]
This is a known problem and these cases will cause problem in Movebank. 
Movebank's suggestion is to split the area into 2 parts. In future they may consider to solve the problem if got time. For now I'll add a check to remind user for these cases.

**Comment by xhdong-umd** [comment]
Yesterday I had some test requests but all got server error later. I tried several times and even tried previous tested request, all had server error. Today these requests are working fine... It's really hard to know where 's the problem when this kind of error happened.

**Comment by xhdong-umd** [comment]
@chfleming , in building package I'm checking on the track mode request too, which will need to upload track csv in a specific format.

With telemetry data object, do you always have time, long/lat in fixed column name like "timestamp", "longitude", "latitude"? I knew as.telemetry will guess them in importing, and I'm wondering if I can assume you always use these 3 name in cleaned up telemetry object.

**Comment by xhdong-umd** [comment]
For most cases, I guess grid mode is always better than track mode? Track mode only give your annotations along the movement track, but grid give you annotations on whole grid area.

If this is the case, I think I will not create convenient functions to submit track request from telemetry data. The basic function is still there so user can submit track request, just they need to specify the parameter used with a little more details, not like the as-simple-as-possible interface in grid mode.

**Comment by xhdong-umd** [comment]
And for extent value of bounding box, does xmin always less than xmax? What happens in anti-meridian case? Will xmin always represent the west side, or the smaller value? I need some way to detect the anti-meridian case.

**Comment by chfleming** [comment]
telemetry objects have fixed column names, yes.

I made xmin the western most point, so it being a larger value would indicate the anti-meridian case.

**Comment by xhdong-umd** [comment]
All functions are reorganized and built into package. Exported functions are all documented.

**Comment by xhdong-umd** [comment]
I'm closing the issue as the first stage objectives are implemented. We can open a new issue for next stage (like projection support)  when needed.
